crates/phactory/src/prpc_service.rs

```diff
@@ -816,4 +825,273 @@ impl<Platform: pal::Platform + Serialize + DeserializeOwned> PhactoryApi
+                    signature: base64::decode(sig).map_err(from_display)?,
+                    signing_cert: base64::decode_config(cert, base64::STANDARD)
+                        .map_err(from_display)?,
+                }),
+                timestamp: now(),
+            });
+        } else {
+            info!("Omit RA report in challenge response in dev mode");
+        }
+
+        Ok(pb::WorkerKeyChallengeResponse {
+            payload: Some(payload),
+            attestation,
+        })
+    }
+
+    fn receive_worker_key(&mut self, request: pb::GetWorkerKeyResponse) -> RpcResult<bool> {
+        let mut phactory = self.lock_phactory();
+        let payload = request
+            .payload
+            .ok_or_else(|| from_display("Encrypted WorkerKey not found"))?;
+        let dev_mode = payload.dev_mode;
+
+        // verify RA report
+        if !dev_mode {
+            let payload_hash = sp_core::hashing::blake2_256(&payload.encode());
+            let raw_attestation = request
+                .attestation
+                .ok_or_else(|| from_display("Attestation not found"))?;
+            let payload = raw_attestation
+                .payload
+                .ok_or_else(|| from_display("Missing attestation payload"))?;
+            let attn_to_validate = Attestation::SgxIas {
+                ra_report: payload.report.as_bytes().to_vec(),
+                signature: payload.signature,
+                raw_signing_cert: payload.signing_cert,
+            };
+            IasValidator::validate(&attn_to_validate, &payload_hash, now(), false, vec![])
+                .map_err(|_| from_display("Invalid RA report"))?;
+        } else {
+            info!("Skip RA report check in dev mode");
+        }
+
+        let encrypted_key = payload.decode_encrypted_key().map_err(from_display)?;
+        if encrypted_key.is_none() {
+            info!("Receive empty key handover");
+            return Ok(false);
+        }
+
+        let encrypted_key = encrypted_key.expect("checked; qed");
+        let my_ecdh_key = phactory
+            .tmp_ecdh_key
+            .as_ref()
+            .ok_or_else(|| from_display("Ecdh key not initialized"))?;
+        let secret = key_share::decrypt_secret_from(
+            my_ecdh_key,
+            &encrypted_key.ecdh_pubkey.0,
+            &encrypted_key.encrypted_key,
+            &encrypted_key.iv,
+        )
+        .map_err(from_debug)?;
+
+        // only seal if the key is successfully updated
+        let genesis_block_hash = payload.decode_genesis_block_hash().map_err(from_display)?;
+        phactory
+            .save_runtime_data(
+                genesis_block_hash,
+                sr25519::Pair::restore_from_secret_key(&secret),
+                false, // we are not sure whether this key is injected
+                dev_mode,
+            )
+            .map_err(from_display)?;
+        // clear cached RA report to prevent replay
+        phactory.runtime_info = None;
```

Comment:
Remove `tmp_ecdh_key`?

Btw, we'd better to call it `handover_ecdh_key`

Reply:
Changed.